### PR TITLE
controller/nodelifecycle: Refactor to not make API calls under lock

### PR DESCRIPTION
**Pull request by MadhavJivrajani**
#### What type of PR is this?

/kind cleanup
/priority important-soon

#### What this PR does / why we need it:

The evictorLock only protects zonePodEvictor and zoneNoExecuteTainter.
processTaintBaseEviction showed indications of increased lock contention
among goroutines (see issue 110341 for more details).

The refactor done is to ensure that all codepaths in that function that
hold the evictorLock AND make API calls under the lock, are now making
API calls outside the lock and the lock is held only for accessing either
zonePodEvictor or zoneNoExecuteTainter or both.

Two other places where the refactor was done is the doEvictionPass and
doNoExecuteTaintingPass functions which make multiple API calls under
the evictorLock.

#### Which issue(s) this PR fixes:

ref https://github.com/kubernetes/kubernetes/issues/110341

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MadhavJivrajani** [comment]
Also from `stress`:

```
go test -race -c && stress ./nodelifecycle.test
5s: 96 runs so far, 0 failures
10s: 200 runs so far, 0 failures
15s: 284 runs so far, 0 failures
20s: 360 runs so far, 0 failures
25s: 417 runs so far, 0 failures
30s: 485 runs so far, 0 failures
35s: 558 runs so far, 0 failures
40s: 632 runs so far, 0 failures
45s: 704 runs so far, 0 failures
50s: 778 runs so far, 0 failures
55s: 851 runs so far, 0 failures
1m0s: 927 runs so far, 0 failures
1m5s: 1005 runs so far, 0 failures
1m10s: 1083 runs so far, 0 failures
```

**Comment by MadhavJivrajani** [comment]
/hold
There are other places where this can be done as well. I can get to them once the approach looks good! Not sure if I'm missing something here.

**Comment by aojea** [comment]
we should have an unit test reproducing the issue, I couldn't check in detail this problem so brainstorming here, but I'm thinking that maybe using fake client go to block on the first call, will make the subsequent executions of the method timeout (never being able to acquire the lock)

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 1]
```diff
@@ -708,11 +708,22 @@ func (nc *Controller) doNoExecuteTaintingPass(ctx context.Context) {
...
-	defer nc.evictorLock.Unlock()
+	zonePodEvictorKeys := make([]string, 0, len(nc.zonePodEvictor))
 	for k := range nc.zonePodEvictor {
+		zonePodEvictorKeys = append(zonePodEvictorKeys, k)
+	}
+	nc.evictorLock.Unlock()
+
+	for _, k := range zonePodEvictorKeys {
+		nc.evictorLock.Lock()
+		zonePodEvictionWorker := nc.zonePodEvictor[k]
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:723]: Shouldn't we check if it still exists in the map.

Also - is scheduler.RateLimitedTimedQueue safe to use without lock?

Comment (MadhavJivrajani) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go:723]: > Shouldn't we check if it still exists in the map.

We should, good point. Thanks!

> is scheduler.RateLimitedTimedQueue safe to use without lock?

From what I understand, it is safe:

https://github.com/MadhavJivrajani/kubernetes/blob/db55f8b80d99ff8eeb14d8985f8461f4756f52b8/pkg/controller/nodelifecycle/scheduler/rate_limited_queue.go#L232-L235

To be sure, I modified `TestTry` to do:

```go
	wg := sync.WaitGroup{}
	wg.Add(10)
	for i := 0; i < 10; i++ {
		go func() {
			defer wg.Done()
			evictor.Try(func(value TimedValue) (bool, time.Duration) {
				deletedMap.Insert(value.Value)
				return true, 0
			})
		}()
	}
	wg.Wait()
```
and:
```
❯ go test -c -race
❯ stress ./scheduler.test -test.run ^TestTry$
5s: 311 runs so far, 0 failures
10s: 627 runs so far, 0 failures
15s: 881 runs so far, 0 failures
20s: 1114 runs so far, 0 failures
25s: 1346 runs so far, 0 failures
30s: 1579 runs so far, 0 failures
35s: 1817 runs so far, 0 failures
40s: 2047 runs so far, 0 failures
45s: 2282 runs so far, 0 failures
50s: 2518 runs so far, 0 failures
55s: 2748 runs so far, 0 failures
1m0s: 2953 runs so far, 0 failures
1m5s: 3190 runs so far, 0 failures
```

Comment (wojtek-t) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go:723]: > Edit: Now that I think about it, I'm not sure how a key would get deleted from the map. I don't see any logic around that, just additions happening here:

OK fair. It's for zones (and zones generally don't get added and removed), which means it's probably OK.

Can you please add a comment to mention say that it's safe?

Comment (wojtek-t) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go:723]: And I agree that zonePodEvictionWorker is thread-safe.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 2]
```diff
@@ -708,11 +708,22 @@ func (nc *Controller) doNoExecuteTaintingPass(ctx context.Context) {
 }
 
 func (nc *Controller) doEvictionPass(ctx context.Context) {
+	// Extract out the keys of the map in order to not hold
+	// the evictorLock for the entire function and hold it
+	// only when nescessary.
 	nc.evictorLock.Lock()
-	defer nc.evictorLock.Unlock()
+	zonePodEvictorKeys := make([]string, 0, len(nc.zonePodEvictor))
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go:715]: ping

**Review by wojtek-t — COMMENTED** [review]
Couple small comments - but overall this looks fine.

**Comment by MadhavJivrajani** [comment]
@wojtek-t done.
A few other functions which were making API calls under the lock are also refactored as part of the force push. PTAL and lmk if things look good! 

Also, what would be the best way to verify that this solves the issue? Will we be triggering the scalability test that detected this or is there an equivalent test that the CI can run?

**Comment by MadhavJivrajani** [comment]
Test failure due to copy-paste error

**Comment by MadhavJivrajani** [comment]
Also:
```
❯ go test -race -c
❯ stress ./nodelifecycle.test
...
6m0s: 10021 runs so far, 0 failures
6m5s: 10135 runs so far, 0 failures
6m10s: 10259 runs so far, 0 failures
```

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 3]
```diff
@@ -664,11 +664,31 @@ func (nc *Controller) doNoScheduleTaintingPass(ctx context.Context, nodeName str
 }
 
 func (nc *Controller) doNoExecuteTaintingPass(ctx context.Context) {
-	nc.evictorLock.Lock()
-	defer nc.evictorLock.Unlock()
-	for k := range nc.zoneNoExecuteTainter {
+	// Extract out the keys of the map in order to not hold
+	// the evictorLock for the entire function and hold it
+	// only when nescessary.
+	zoneNoExecuteTainterKeys := make([]string, 0, len(nc.zoneNoExecuteTainter))
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:670]: unsafe access to nc.zoneNoExecuteTainer

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 4]
```diff
@@ -708,11 +728,33 @@ func (nc *Controller) doNoExecuteTaintingPass(ctx context.Context) {
 }
 
 func (nc *Controller) doEvictionPass(ctx context.Context) {
-	nc.evictorLock.Lock()
-	defer nc.evictorLock.Unlock()
-	for k := range nc.zonePodEvictor {
+	// Extract out the keys of the map in order to not hold
+	// the evictorLock for the entire function and hold it
+	// only when nescessary.
+	zonePodEvictorKeys := make([]string, 0, len(nc.zonePodEvictor))
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:734]: unsafe access to nc.zoneNoExecuteTainer

**Review by wojtek-t — COMMENTED** [review]
Two comments - other than that LGTM.

**Comment by MadhavJivrajani** [comment]
@wojtek-t good catch, thank!
I've made the changes.

**Comment by MadhavJivrajani** [comment]
The second force push removed the TODO:
https://github.com/kubernetes/kubernetes/blob/451e1fa8bcff38b87504eebd414948e505526d01/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L281

**Comment by wojtek-t** [comment]
/lgtm
/approve

/hold cancel

@marseel @mborsz - FYI

**Pull request merged**